### PR TITLE
Rk multiroom merge

### PR DESCRIPTION
**Pull request by krzysieksr**
multi-room

### CODE REVIEW

File: backend/src/main/java/pl/krkteam/battleships/room/holder/Room.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            return new OpponentPresentDTO();
+        }
+        return new OpponentAbsentDTO();
+    }
+
+    private void initializeGame() {
+        if (!isGameInitialized) {
+            List<Player> playerList = new ArrayList<>(playerSet);
+            game.initializeGame(playerList.get(0), playerList.get(1));
+            isGameInitialized=true;
```
Comment (jeger) [inline backend/src/main/java/pl/krkteam/battleships/room/holder/Room.java]: isGameInitialized = true

File: backend/src/main/java/pl/krkteam/battleships/room/holder/RoomHolder.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        final Room room = roomMap.get(roomName);
+        if (room.isPlayerBelongToRoom(player)) {
+            return new OpponentAbsentDTO();
+        }
+        return room.areBothPlayers();
+    }
+
+    public Room getRoom(String roomName) {
+        final Room room = roomMap.get(roomName);
+        if (null == room) {
```
Comment (jeger) [inline backend/src/main/java/pl/krkteam/battleships/room/holder/RoomHolder.java]: czemu od tyłu

File: frontend/src/app/services/drag-ship/drag-ship.service.ts [hunk 3]
```diff
@@ -8,71 +8,48 @@ export class DragShipService {
...
+      // let row = this.getRow(boardDiv, rowIndex);
+      // let cell = this.getCell(row, rowIndex, columnIndex);
+      //
+      // let innerColor = this.getInnerColor(cell);
+      //
+      //   if (this.isThereNoShip(cell)) {
+      //
+      //     if (innerColor != null) {
+      //     innerColor.classList.add("inner-color-black");
+      //
```
Comment (jeger) [inline frontend/src/app/services/drag-ship/drag-ship.service.ts]: commented code

File: frontend/src/app/views/boards/player-board/player-board.component.ts [hunk 4]
```diff
@@ -29,41 +30,44 @@ export class PlayerBoardComponent implements OnInit {
...
+  //   this.playerBoard.generateBoardWithWater(10);
+  //
+  //   this.putShipsFromShipGenerator_random();
+  // }
+  //
+  // putShipsFromShipGenerator_random() {
+  //   const shipArray = this.shipGenerator.generateShipsRandomly(this.playerBoard);
+  //
+  //   this.testShipSending(shipArray);
+  // }
```
Comment (jeger) [inline frontend/src/app/views/boards/player-board/player-board.component.ts]: commented code

**Review by jeger — CHANGES_REQUESTED** [review]

File: backend/src/main/java/pl/krkteam/battleships/room/holder/RoomHolder.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            throw new NoSuchElementException();
+        }
+        return room;
+    }
+
+
```
Comment (JaroslawZaczyk) [inline backend/src/main/java/pl/krkteam/battleships/room/holder/RoomHolder.java]: empty line

File: backend/src/main/java/pl/krkteam/battleships/room/holder/converters/RoomHolderToRoomListDTO.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.core.convert.converter.Converter;
+import org.springframework.stereotype.Component;
+import pl.krkteam.battleships.room.holder.Room;
+import pl.krkteam.battleships.room.holder.RoomHolder;
+import pl.krkteam.battleships.room.holder.dto.room.list.RoomDTO;
+import pl.krkteam.battleships.room.holder.dto.room.list.RoomListDTO;
+
+import java.util.ArrayList;
+import java.util.List;
+
```
Comment (JaroslawZaczyk) [inline backend/src/main/java/pl/krkteam/battleships/room/holder/converters/RoomHolderToRoomListDTO.java:12]: empty line

File: backend/src/main/java/pl/krkteam/battleships/room/holder/dto/room/list/RoomListDTO.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
+package pl.krkteam.battleships.room.holder.dto.room.list;
+
+import java.util.Arrays;
+
+public class RoomListDTO {
+    private RoomDTO[] roomName;
```
Comment (JaroslawZaczyk) [inline backend/src/main/java/pl/krkteam/battleships/room/holder/dto/room/list/RoomListDTO.java:6]: try if this field can be a list

File: backend/src/main/java/pl/krkteam/battleships/shooting/controller/ShootingController.java [hunk 8]
```diff
@@ -19,34 +21,42 @@
...
-        this.game = game;
+        this.roomHolder = roomHolder;
     }
 
     @PostMapping(value = "/game/player/shot")
-    public String validateShot(@RequestBody String shotJson, @RequestParam String playerName) {
+    public String validateShot(@RequestBody String shotJson, @RequestParam String playerName,
+                               @RequestParam String roomName) {
         final Player shootingPlayer = new Player(playerName);
         Gson gson = new Gson();
```
Comment (JaroslawZaczyk) [inline backend/src/main/java/pl/krkteam/battleships/shooting/controller/ShootingController.java:35]: check if you have to create a new GSON each time when you send sth to controller

File: backend/src/main/java/pl/krkteam/battleships/shooting/controller/ShootingController.java [hunk 9]
```diff
@@ -19,34 +21,42 @@
...
 
+        final Game game = getGameForPlayer(roomName);
```
Comment (JaroslawZaczyk) [inline backend/src/main/java/pl/krkteam/battleships/shooting/controller/ShootingController.java]: e.g. getGameFromRoom

File: backend/src/main/java/pl/krkteam/battleships/shooting/controller/ShootingController.java [hunk 10]
```diff
@@ -19,34 +21,42 @@
...
+        sendResponseToOpponentPlayer(shotResultDTO, shotDTO, opponentPlayer, game);
 
         calculateTurn(shootingPlayer, shotResultDTO, turnHolder);
 
         return gson.toJson(shotResultDTO);
     }
 
+    private Game getGameForPlayer(String roomName) {
+        final Room room = roomHolder.getRoom(roomName);
+        return room.getGame();
```
Comment (JaroslawZaczyk) [inline backend/src/main/java/pl/krkteam/battleships/shooting/controller/ShootingController.java]: I think you can inline this method up - Artem said.

File: backend/src/main/java/pl/krkteam/battleships/wait/opponent/WaitForOpponentController.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+    @PostMapping(value = "room/opponent/present")
+    public String isOpponentPresent(@RequestParam String playerName, @RequestParam String roomName) {
+
+        Gson gson = new Gson();
+
+        Player waitingPlayer = new Player(playerName);
+
+        final WaiterResponseDTO waiterResponseDTO = roomHolder.isOpponentInRoom(roomName, waitingPlayer);
+
```
Comment (JaroslawZaczyk) [inline backend/src/main/java/pl/krkteam/battleships/wait/opponent/WaitForOpponentController.java]: empty lines

**Review by jeger — APPROVED** [review]

File: backend/src/main/java/pl/krkteam/battleships/room/holder/dto/room/list/RoomListDTO.java (hunk 7 above)
Comment (krzysieksr) [reply backend/src/main/java/pl/krkteam/battleships/room/holder/dto/room/list/RoomListDTO.java:6]: For sure transfered object must be in array (in angular there is no List) and in Java you can use Arrays.asList() or use TypeToken to transfer array to list.

File: backend/src/main/java/pl/krkteam/battleships/room/holder/RoomsController.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        final RoomListDTO roomListDTO = roomHolderToRoomListDTO.convert(roomHolder);
+
+        Gson gson = new Gson();
+        return gson.toJson(roomListDTO);
+    }
+
+    @PostMapping(value = "room/join")
+    public String joinRoom(@RequestParam String playerName, @RequestParam String roomName) {
+        Gson gson = new Gson();
+
```
Comment (JaroslawZaczyk) [inline backend/src/main/java/pl/krkteam/battleships/room/holder/RoomsController.java:38]: empty lines

**Review by JaroslawZaczyk — APPROVED** [review]

**Review by RafalJDev — APPROVED** [review]
I did review this PR

**Pull request merged**